ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java

```diff
@@ -86,28 +89,29 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, NetworkMetr
    * Attempt to send the given requests and poll for responses from the network via the associated selector. Any
    * requests that could not be sent out will be added to a queue. Every time this method is called, it will first
    * attempt sending the requests in the queue (or time them out) and then attempt sending the newly added requests.
-   * @param requestInfos the list of {@link RequestInfo} representing the requests that need to be sent out. This
+   * @param requestsToSend the list of {@link RequestInfo} representing the requests that need to be sent out. This
    *                     could be empty.
+   * @param requestsToDrop the list of correlation IDs representing the requests that can be dropped by
+   *                       closing the connection.
    * @param pollTimeoutMs the poll timeout.
    * @return a list of {@link ResponseInfo} representing the responses received for any requests that were sent out
    * so far.
    * @throws IllegalStateException if the NetworkClient is closed.
    */
-  public List<ResponseInfo> sendAndPoll(List<RequestInfo> requestInfos, int pollTimeoutMs) {
+  public List<ResponseInfo> sendAndPoll(List<RequestInfo> requestsToSend, Set<Integer> requestsToDrop,
+      int pollTimeoutMs) {
     if (closed || !selector.isOpen()) {
       throw new IllegalStateException("The NetworkClient is closed.");
     }
     long startTime = time.milliseconds();
     List<ResponseInfo> responseInfoList = new ArrayList<>();
     try {
-      for (RequestInfo requestInfo : requestInfos) {
+      for (RequestInfo requestInfo : requestsToSend) {
         pendingRequests.add(new RequestMetadata(requestInfo));
       }
-      List<NetworkSend> sends = prepareSends(responseInfoList);
-      if (networkConfig.networkClientEnableConnectionReplenishment) {
-        int connectionsInitiated = connectionTracker.replenishConnections(this::connect);
-        networkMetrics.connectionReplenished.inc(connectionsInitiated);
-      }
+      List<NetworkSend> sends = prepareSends(requestsToDrop, responseInfoList);
+      replenishConnections();
+      dropInFlightRequests(requestsToDrop);
```

Comment:
No. I mean NonBlockingRouter.run(). 

But current solution also looks good to me.

Reply:
I think it is better to keep it in NetworkClient since there are some ordering constraints and the difference in how pending requests and in flight requests are handled.